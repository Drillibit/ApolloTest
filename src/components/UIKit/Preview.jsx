import React, { PureComponent } from 'react';
import styled, { css, keyframes } from 'styled-components';
import PropTypes from 'prop-types';

import { H3, SmallText } from './Typography';
import { colors } from '../helpers/colors';
import { Rating } from './Rating';
import { Button, StyledButton } from './Button';
import { Icon } from './Icon';
import { FavouriteButton } from '../FavouriteButton';

const StyledCustomBtn = styled(StyledButton)`
  padding: 4px 43px;
`;

const StyledParent = styled.div`
  position: relative;
  width: 282px;
  height: 284px;
`;

const StyledPreviewContainerOpen = css`
  z-index: 10;
  right: -67px;
  width: 412px;
  height: 526px;
  border-radius: 2px;
  background-color: #ffffff;
  box-shadow: 0 12px 75px 2px rgba(0, 0, 0, 0.41),
    0 2px 9px 1px rgba(0, 0, 0, 0.28);
`;

const StyledPreviewContainer = styled.div`
  position: absolute;
  overflow: hidden;
  width: 282px;
  height: 284px;
  border-radius: 2px;
  background-color: #ffffff;
  box-shadow: 0 2px 7px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  transition: all ease 0.3s;
  z-index: 0;
  ${({ open }) => (open ? StyledPreviewContainerOpen : '')};
`;

const StyledHeaderClose = css`
  transform: translateX(-500px);
  display: none;
`;

const StyledHeader = H3.extend`
  color: ${colors.grey500};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 16px 12px 2px 12px;
  transition: transform ease .3s;
  ${({ open }) => (open ? StyledHeaderClose : '')}
  display: ${setTimeout(() => 'none', 1000)};
`;

const BgAnimation = keyframes`
  0% {height: 70%},
  25% {height: 65%},
  50% {height: 60%},
  75% {height: 55%},
  100% {height: 50%}
`;
const BgKeeperMove = css`
  transform: translateY(-10%);
  min-height: 50%;
  min-height: 250px;
  animation: ${BgAnimation} 0.3s ease-in;
`;

const BgKeeper = styled.div`
  width: 100%;
  flex: 2;
  border-radius: 0 0 2px 2px;
  transition: all ease-in-out .4s;
  background-image: url('${({ bg }) => bg}');
  background-size: cover;
  ${({ open }) => (open ? BgKeeperMove : '')}
`;

const RatingMove = css`
  transform: translateX(600px);
`;

const RatingContainer = styled.div`
  transition: transform ease 0.3s;
  margin-bottom: 11px;
  ${({ open }) => (open ? RatingMove : '')};
`;

const StyledInfoContainerMove = css`
  max-height: 256px;
  display: flex;
  flex-direction: column;
`;

const StyledInfoContainer = styled.div`
  display: none;
  transform: all ease 0.3s;
  padding: 0 16px 14px 16px;
  transition: height 0.3s linear;
  overflow: hidden;
  min-width: 412px;
  ${({ open }) => (open ? StyledInfoContainerMove : '')};
`;

const StyledHeaderInfo = H3.extend`
  color: ${colors.grey500};
  width: 100%;
`;

const StyledSmallInfo = styled.span`
  color: ${colors.grey200};
  font-size: 14px;
  margin-right: 10px;
`;

const StyledDigitContainer = styled.div`
  display: flex;
`;

const StyledDetails = styled.div`
  display: flex;
  margin: 7px 0;
  align-items: baseline;
`;

const StyledDetailsText = SmallText.extend`
  color: ${colors.grey500};
`;

const StyledDetailsHeader = StyledSmallInfo.extend`
  min-width: 60px;
`;

const StyledParagraph = StyledDetailsText.extend`
  max-height: 38px;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 0 auto;
`;

export class Preview extends PureComponent {
  static propTypes = {
    description: PropTypes.string,
    title: PropTypes.string,
    bg: PropTypes.string,
    year: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    duration: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    pg: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    genre: PropTypes.string,
    cast: PropTypes.string
  };

  static defaultProps = {
    description: '',
    title: '',
    bg: '',
    duration: '',
    year: '',
    pg: '',
    genre: '',
    cast: ''
  };

  state = {
    inOpenState: false
  };

  handleDisplay = () => {
    this.timeOut = setTimeout(() => {
      this.setState({
        inOpenState: true
      });
    }, 1000);
  };

  handleHide = () => {
    this.setState({
      inOpenState: false
    });
    clearTimeout(this.timeOut);
  };

  render() {
    const { inOpenState } = this.state;
    const {
      description,
      title,
      year,
      bg,
      duration,
      pg,
      genre,
      cast,
      isFavourite,
      toggleFavourite,
    } = this.props;

    console.log(this.props);

    return (
      <StyledParent>
        <StyledPreviewContainer
          onMouseEnter={this.handleDisplay}
          onMouseLeave={this.handleHide}
          open={inOpenState}
        >
          <StyledHeader open={inOpenState}>{title}</StyledHeader>
          <RatingContainer open={inOpenState}>
            <Rating {...this.props} />
          </RatingContainer>
          <BgKeeper open={inOpenState} bg={bg} />
          <StyledInfoContainer open={inOpenState}>
            <StyledHeaderInfo>{title}</StyledHeaderInfo>
            <StyledDigitContainer>
              <StyledSmallInfo>{year}</StyledSmallInfo>
              <StyledSmallInfo>
                {duration} {duration && 'мин'}
              </StyledSmallInfo>
              <StyledSmallInfo>{pg}</StyledSmallInfo>
            </StyledDigitContainer>
            {genre && (
              <StyledDetails>
                <StyledDetailsHeader>Жанр:</StyledDetailsHeader>
                <StyledDetailsText>{genre}</StyledDetailsText>
              </StyledDetails>
            )}
            {cast && (
              <StyledDetails>
                <StyledDetailsHeader>В ролях:</StyledDetailsHeader>
                <StyledDetailsText>{cast}</StyledDetailsText>
              </StyledDetails>
            )}
            <StyledDetails>
              <StyledParagraph>
                {description.length > 90
                  ? `${description.slice(0, 90)}...`
                  : description}
              </StyledParagraph>
            </StyledDetails>
            <ButtonContainer>
              <FavouriteButton
                btnType="transparent-dark"
                isFavourite={isFavourite}
                onClick={toggleFavourite}
              />
              <StyledCustomBtn btnType="primary" onClick={this.handleHide}>
                Подробнее
              </StyledCustomBtn>
            </ButtonContainer>
          </StyledInfoContainer>
        </StyledPreviewContainer>
      </StyledParent>
    );
  }
}
