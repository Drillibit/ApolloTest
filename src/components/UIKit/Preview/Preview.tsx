import React, { PureComponent } from 'react';
import { Link } from 'react-router-dom';
import styled, { css, keyframes } from 'styled-components';
import RootClose from 'react-overlays/lib/RootCloseWrapper';
import { Query } from 'react-apollo';

import { Icon } from '../Icon';
import { H3, SmallText } from '../Typography';
import { colors } from '../../helpers/colors';
import { Rating } from '../Rating';
import { StyledButton } from '../Button';
import { FavouriteControll } from '../../../containers/FavouriteControll';
import { GET_GENRES } from '../../Requests/frontPage';

type StyledPreviewType = {
  open: boolean
  bg?: string
};

const StyledCustomBtn = styled(StyledButton)`
  padding: 4px 43px;
`;

const StyledParent = styled.div`
  position: relative;
  width: 282px;
  height: 284px;
  margin-bottom: 10px;
`;

const StyledPreviewContainerOpen = css`
  z-index: 10;
  right: -67px;
  width: 430px;
  height: 526px;
  border-radius: 2px;
  background-color: #ffffff;
  box-shadow: 0 12px 75px 2px rgba(0, 0, 0, 0.41),
    0 2px 9px 1px rgba(0, 0, 0, 0.28);
`;

const StyledPreviewContainer = styled.div<StyledPreviewType>`
  position: absolute;
  overflow: hidden;
  width: 282px;
  height: 284px;
  border-radius: 2px;
  background-color: #ffffff;
  box-shadow: 0 2px 7px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  transition: all ease 0.3s;
  z-index: 0;
  ${({ open }) => (open ? StyledPreviewContainerOpen : '')};
`;

const StyledHeaderClose = css`
  transform: translateX(-500px);
  display: none;
`;

const StyledHeader = styled(H3)<StyledPreviewType>`
  color: ${colors.grey500};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 16px 12px 2px 12px;
  transition: transform ease 0.3s;
  ${({ open }) => (open ? StyledHeaderClose : '')};
`;

const BgAnimation = keyframes`
  0% {height: 70%},
  25% {height: 65%},
  50% {height: 60%},
  75% {height: 55%},
  100% {height: 50%}
`;
const BgKeeperMove = css`
  transform: translateY(-14%);
  min-height: 50%;
  min-height: 250px;
  animation: ${BgAnimation} 0.3s ease-in;
`;

const BgKeeper = styled.div<StyledPreviewType>`
  width: 100%;
  flex: 2;
  border-radius: 0 0 2px 2px;
  transition: all ease-in-out .4s;
  background-image: url('${({ bg }) => bg}');
  background-size: cover;
  ${({ open }) => (open ? BgKeeperMove : '')}
`;

const RatingMove = css`
  transform: translateX(600px);
`;

const RatingContainer = styled.div<StyledPreviewType>`
  transition: transform ease 0.3s;
  margin-bottom: 11px;
  ${({ open }) => (open ? RatingMove : '')};
`;

const StyledInfoContainerMove = css`
  max-height: 256px;
  display: flex;
  flex-direction: column;
`;

const StyledInfoContainer = styled.div<StyledPreviewType>`
  display: none;
  transform: all ease 0.3s;
  padding: 0 16px 14px 16px;
  transition: height 0.3s linear;
  overflow: hidden;
  min-width: 412px;
  ${({ open }) => (open ? StyledInfoContainerMove : '')};
`;

const StyledHeaderInfo = styled(H3)`
  color: ${colors.grey500};
  width: 100%;
`;

const StyledSmallInfo = styled.span`
  color: ${colors.grey200};
  font-size: 14px;
  margin-right: 10px;
`;

const StyledDigitContainer = styled.div`
  display: flex;
`;

const StyledDetails = styled.div`
  display: flex;
  margin: 7px 0;
  align-items: baseline;
`;

const StyledDetailsText = styled(SmallText)`
  color: ${colors.grey500};
`;

const StyledDetailsHeader = styled(StyledSmallInfo)`
  min-width: 43px;
  margin-right: 0;
`;

const StyledParagraph = styled(StyledDetailsText)`
  max-height: 38px;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 0 auto;
`;

export interface PreviewProp {
  id: string
  description: string
  title: string
  bg: string
  year: string | number
  duration: string | number
  pg: string | number
  genre_ids: string[]
  cast?: string
  isFavourite: boolean,
  toggleFavourite?: () => void
};
export class Preview extends PureComponent<PreviewProp> {
  static defaultProps = {
    id: '0',
    description: '',
    title: '',
    bg: '',
    duration: '',
    year: '',
    pg: '',
    genre_ids: '',
    cast: '',
    isFavourite: false,
  };

  state = {
    inOpenState: false
  };

  handleDisplay = () => {
    this.setState({
      inOpenState: !this.state.inOpenState
    });
  };

  handleHide = () => {
    this.setState({
      inOpenState: false
    });
  };

  render() {
    const { inOpenState } = this.state;
    const {
      id,
      description,
      title,
      year,
      bg,
      duration,
      pg,
      genre_ids,
      cast
    } = this.props;

    return (
      <RootClose onRootClose={this.handleHide}>
        <StyledParent>
          <StyledPreviewContainer
            onClick={this.handleDisplay}
            open={inOpenState}
          >
            <StyledHeader open={inOpenState}>{title}</StyledHeader>
            <RatingContainer open={inOpenState}>
              <Rating {...this.props} />
            </RatingContainer>
            <BgKeeper open={inOpenState} bg={bg} />
            <StyledInfoContainer open={inOpenState}>
              <StyledHeaderInfo>{title}</StyledHeaderInfo>
              <StyledDigitContainer>
                <StyledSmallInfo>{year}</StyledSmallInfo>
                <StyledSmallInfo>
                  {duration} {duration && 'мин'}
                </StyledSmallInfo>
                <StyledSmallInfo>{pg}</StyledSmallInfo>
              </StyledDigitContainer>
              <Query query={GET_GENRES}>
                {({ error, loading, data: { genres_arr } }) => {
                  if (error) return `Error: ${error.message}`;
                  if (loading) return 'Loading...';
                  if (genres_arr) {
                    const genreObj = genres_arr.reduce((acc:any, item:any) => ({
                      ...acc,
                      [item.id]: { ...item }
                    }));
                    return (
                      <StyledDetails>
                        <StyledDetailsHeader>Жанр:</StyledDetailsHeader>
                        <StyledDetailsText>
                          {genre_ids.length &&
                            genre_ids.map((gen) => {
                              if (genreObj[gen]) {
                                if (genreObj[gen].name) {
                                  return `${genreObj[gen].name} `;
                                }
                              }
                              return null;
                            })}
                        </StyledDetailsText>
                      </StyledDetails>
                    );
                  }
                }}
              </Query>
              {cast && (
                <StyledDetails>
                  <StyledDetailsHeader>В ролях:</StyledDetailsHeader>
                  <StyledDetailsText>{cast}</StyledDetailsText>
                </StyledDetails>
              )}
              <StyledDetails>
                <StyledParagraph>
                  {description.length > 90
                    ? `${description.slice(0, 90)}...`
                    : description}
                </StyledParagraph>
              </StyledDetails>
              <ButtonContainer>
                <FavouriteControll
                  btnType="transparent-dark"
                  btnSize="small"
                  movieId={id}
                >
                  <Icon size="sm" color="#fff" icon="heart" />
                  Избранное
                </FavouriteControll>
                <Link to={`/movie/${id}`}>
                  <StyledCustomBtn btnType="primary" onClick={this.handleHide}>
                    Подробнее
                  </StyledCustomBtn>
                </Link>
              </ButtonContainer>
            </StyledInfoContainer>
          </StyledPreviewContainer>
        </StyledParent>
      </RootClose>
    );
  }
}
