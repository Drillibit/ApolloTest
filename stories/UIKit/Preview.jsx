import { storiesOf } from '@storybook/react';
import React, { Fragment } from 'react';

import { Container } from '../helpers/Container';
import { Preview } from '../../src/components/UIKit/Preview';

const stories = storiesOf('UIKit/Preview', module);

const propsOne = {
  voteCount: 2079,
  id: 351286,
  video: false,
  voteAverage: 6.6,
  title: 'Jurassic World: Fallen Kingdom',
  bg: 'https://image.tmdb.org/t/p/w500/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg',
  original_title: 'Jurassic World: Fallen Kingdom',
  genre_ids: [
    28,
    12,
    878
  ],
  adult: false,
  description: 'Several years after the demise of Jurassic World, a volcanic eruption threatens the remaining dinosaurs on the island of Isla Nublar. Claire Dearing, the former park manager and founder of the Dinosaur Protection Group, recruits Owen Grady to help prevent the extinction of the dinosaurs once again.',
  yaer: '2018-06-06'
};

const propsTwo = {
  voteCount: 2079,
  id: 351286,
  video: false,
  voteAverage: 6.6,
  title: 'Jurassic World: Fallen Kingdom',
  bg: 'https://image.tmdb.org/t/p/w500/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg',
  original_title: 'Jurassic World: Fallen Kingdom',
  genre_ids: [
    28,
    12,
    878
  ],
  adult: false,
  description: 'Several years after the demise of Jurassic World, a volcanic eruption threatens the remaining dinosaurs on the island of Isla Nublar. Claire Dearing, the former park manager and founder of the Dinosaur Protection Group, recruits Owen Grady to help prevent the extinction of the dinosaurs once again.',
  yaer: '2018-06-06'
};


stories.addWithJSX(
  'Preview component',
  () => (
    <Container>
      <Preview {...propsOne} />
    </Container>
  )
);
