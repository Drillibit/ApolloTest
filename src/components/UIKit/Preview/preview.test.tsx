import React from 'react';
import { Preview, PreviewProp } from './Preview'
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

const fakeMovieData:PreviewProp = {
    id: '351286',
    title: 'Jurassic World: Fallen Kingdom',
    bg: 'https://image.tmdb.org/t/p/w500/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg',
    genre_ids: ['28', '12', '878'],
    isFavourite: false,
    pg: "16",
    duration: '123',
    description:
      'Several years after the demise of Jurassic World, a volcanic eruption threatens the remaining dinosaurs on the island of Isla Nublar. Claire Dearing, the former park manager and founder of the Dinosaur Protection Group, recruits Owen Grady to help prevent the extinction of the dinosaurs once again.',
    year: '2018-06-06',
};


describe('<Preview />', () => {
    it('renders preview component', () => {
        const wrapper = shallow(<Preview {...fakeMovieData} />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    })
});
