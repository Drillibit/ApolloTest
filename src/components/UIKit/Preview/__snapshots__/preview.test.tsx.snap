// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Preview /> renders preview component 1`] = `
<RootCloseWrapper
  event="click"
  onRootClose={[Function]}
>
  <styled.div>
    <styled.div
      onClick={[Function]}
      open={false}
    >
      <Styled(styled.h3)
        open={false}
      >
        Jurassic World: Fallen Kingdom
      </Styled(styled.h3)>
      <styled.div
        open={false}
      >
        <Component
          bg="https://image.tmdb.org/t/p/w500/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg"
          cast=""
          className=""
          description="Several years after the demise of Jurassic World, a volcanic eruption threatens the remaining dinosaurs on the island of Isla Nublar. Claire Dearing, the former park manager and founder of the Dinosaur Protection Group, recruits Owen Grady to help prevent the extinction of the dinosaurs once again."
          duration="123"
          genre_ids={
            Array [
              "28",
              "12",
              "878",
            ]
          }
          id="351286"
          isFavourite={false}
          pg="16"
          size="md"
          title="Jurassic World: Fallen Kingdom"
          voteAverage={0}
          voteCount={0}
          year="2018-06-06"
        />
      </styled.div>
      <styled.div
        bg="https://image.tmdb.org/t/p/w500/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg"
        open={false}
      />
      <styled.div
        open={false}
      >
        <Styled(styled.h3)>
          Jurassic World: Fallen Kingdom
        </Styled(styled.h3)>
        <styled.div>
          <styled.span>
            2018-06-06
          </styled.span>
          <styled.span>
            123
             
            мин
          </styled.span>
          <styled.span>
            16
          </styled.span>
        </styled.div>
        <Query
          query={
            Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": Object {
                    "kind": "Name",
                    "value": "genres",
                  },
                  "operation": "query",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "genres_arr",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 65,
                "start": 0,
              },
            }
          }
        >
          <Component />
        </Query>
        <styled.div>
          <Styled(Styled(Styled(styled.p)))>
            Several years after the demise of Jurassic World, a volcanic eruption threatens the remain...
          </Styled(Styled(Styled(styled.p)))>
        </styled.div>
        <styled.div>
          <Apollo(Apollo(Component))
            btnSize="small"
            btnType="transparent-dark"
            movieId="351286"
          >
            <Icon
              color="#fff"
              icon="heart"
              size="sm"
            />
            Избранное
          </Apollo(Apollo(Component))>
          <Link
            replace={false}
            to="/movie/351286"
          >
            <Styled(styled.button)
              btnType="primary"
              onClick={[Function]}
            >
              Подробнее
            </Styled(styled.button)>
          </Link>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
</RootCloseWrapper>
`;
